File: BaekjoonOnlineJudge/2216/2216.py
import sys
input = sys.stdin.readline


def DFS(s1, s2):
    if s1 >= len(X):
        return B * (len(Y) - s2)
    if s2 >= len(Y):
        return B * (len(X) - s1)

    if memoization[s1][s2] != -2147483647:
        return memoization[s1][s2]
    
    if X[s1] == Y[s2]:
        memoization[s1][s2] = max(memoization[s1][s2], A + DFS(s1 + 1, s2 + 1))
    else:
        memoization[s1][s2] = max(memoization[s1][s2], C + DFS(s1 + 1, s2 + 1))

    memoization[s1][s2] = max(memoization[s1][s2], B + DFS(s1 + 1, s2), B + DFS(s1, s2 + 1))
     
    return memoization[s1][s2]


A, B, C = map(int, input().split())
X = input().rstrip()
Y = input().rstrip()

memoization = [[-2147483647] * len(Y) for _ in range(len(X))]

print(DFS(0, 0))

File: BaekjoonOnlineJudge/2216/2216.cpp
#include <iostream>
#include <algorithm>
#include <string>

using namespace std;

const int INF = -2147483647;
int A, B, C;
string X, Y;
int memoization[3001][3001];

int DFS(int s1, int s2) {
	if (s1 >= X.size())
		return B * (Y.size() - s2);
	if (s2 >= Y.size())
		return B * (X.size() - s1);

	int &score = memoization[s1][s2];
	if (score != INF)
		return score;
	
	if (X[s1] == Y[s2])
		score = max(score, A + DFS(s1 + 1, s2 + 1));
	else
		score = max(score, C + DFS(s1 + 1, s2 + 1));

	score = max(score, max(B + DFS(s1 + 1, s2), B + DFS(s1, s2 + 1)));

	return score;
}

int main(void) {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);

	cin >> A >> B >> C;	
	cin >> X;
	cin >> Y;

	fill(&memoization[0][0], &memoization[X.size() - 1][Y.size()],INF);

	cout << DFS(0, 0) << "\n";
}